### PR TITLE
feature/web

### PR DESCRIPTION
**Pull request by AlekseySayarkin**

### CODE REVIEW

File: web/src/main/java/com.epam.esm/config/RootConfig.java [hunk 1]
```diff
@@ -30,7 +32,7 @@
 @Configuration
 @PropertySource("classpath:db.properties")
 @EnableTransactionManagement
-public class AppConfig {
+public class RootConfig {
```
Comment (churaalena) [inline web/src/main/java/com.epam.esm/config/RootConfig.java:35]: Please rename it

File: web/src/main/java/com.epam.esm/config/WebConfig.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    @Override
+    public void addResourceHandlers(ResourceHandlerRegistry registry) {
+        registry.addResourceHandler("/css/**").addResourceLocations("/css/");
+    }
+
+    @Bean
+    public InternalResourceViewResolver resolver() {
+        InternalResourceViewResolver resolver = new InternalResourceViewResolver();
+        resolver.setPrefix("/WEB-INF/view/");
+        resolver.setSuffix(".jsp");
```
Comment (churaalena) [inline web/src/main/java/com.epam.esm/config/WebConfig.java:25]: Why do you need jsp?

File: web/src/main/java/com.epam.esm/controller/TagController.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    private final TagService tagService;
+
+    @Autowired
+    public TagController(TagService tagService) {
+        this.tagService = tagService;
+    }
+
+    @GetMapping("/tags")
+    public String home(Model model) throws ServiceException {
```
Comment (churaalena) [inline web/src/main/java/com.epam.esm/controller/TagController.java:22]: Please rename it

File: web/src/main/java/com.epam.esm/config/RootConfig.java (hunk 1 above)
Comment (AlekseySayarkin) [reply web/src/main/java/com.epam.esm/config/RootConfig.java]: done

File: web/src/main/java/com.epam.esm/config/WebConfig.java (hunk 2 above)
Comment (AlekseySayarkin) [reply web/src/main/java/com.epam.esm/config/WebConfig.java]: removed jsp

File: web/src/main/java/com.epam.esm/controller/TagController.java (hunk 3 above)
Comment (AlekseySayarkin) [reply web/src/main/java/com.epam.esm/controller/TagController.java]: done

File: service/src/main/java/com/epam/esm/service/exception/ServiceException.java [hunk 4]
```diff
@@ -1,7 +1,9 @@
 package com.epam.esm.service.exception;
 
 import com.epam.esm.dao.exception.ErrorCode;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
 
+@JsonIgnoreProperties(value = {"stackTrace", "cause", "suppressed", "localizedMessage"}, allowSetters = true)
```
Comment (churaalena) [inline service/src/main/java/com/epam/esm/service/exception/ServiceException.java:6]: Why it was added?

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 5]
```diff
@@ -17,15 +22,19 @@
 
 public class GiftCertificateServiceImpl implements GiftCertificateService {
 
+    private enum SortParameter {
```
Comment (churaalena) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:25]: Why its a private enum?

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 6]
```diff
@@ -35,26 +44,30 @@ public GiftCertificateServiceImpl(GiftCertificateDAO giftCertificateDAO, TagDao
     @Override
     public GiftCertificate getGiftCertificate(String name) throws ServiceException {
         if (name == null || name.isEmpty()) {
-            throw new ServiceException("Invalid name", new ErrorCode(INVALID_INPUT));
+            throw new ServiceException("Invalid name", new ErrorCode(ErrorCodeEnum.INVALID_INPUT.getCode()));
```
Comment (churaalena) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:47]: Can we pass just an enum value?

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 7]
```diff
@@ -63,33 +76,39 @@ public GiftCertificate getGiftCertificate(int id) throws ServiceException {
...
 
     @Override
     public List<GiftCertificate> getAllGiftCertificates(String content) throws ServiceException {
         if (content == null || content.isEmpty()) {
-            throw new ServiceException("Invalid content",  new ErrorCode(INVALID_INPUT));
+            throw new ServiceException("Invalid content",  new ErrorCode(ErrorCodeEnum.INVALID_INPUT.getCode()));
         }
 
         try {
             return giftCertificateDAO.getAllGiftCertificates(content);
```
Comment (churaalena) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:92]: Can we rename this method?

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 8]
```diff
@@ -109,66 +130,119 @@ public GiftCertificate getGiftCertificate(int id) throws ServiceException {
...
+            return getSortedCertificates(requestBody.getSort());
+        }
+        if (requestBody.getTagName() != null) {
+            return getGiftCertificateByTagName(requestBody.getTagName());
         }
+        throw new ServiceException("Wrong request input", new ErrorCode(ErrorCodeEnum.INVALID_SORT_INPUT.getCode()));
+    }
+
+    private List<GiftCertificate> getSortedCertificates(String sort) throws ServiceException {
+        int separator = sort.indexOf(".");
```
Comment (churaalena) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:153]: Please refactor this method not to use params with dot

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 9]
```diff
@@ -109,66 +130,119 @@ public GiftCertificate getGiftCertificate(int id) throws ServiceException {
...
     @Override
     @Transactional
     public int addGiftCertificate(GiftCertificate giftCertificate) throws ServiceException {
-        if (CertificateValidator.isNonValid(giftCertificate)) {
-            throw new ServiceException("Invalid certificate",  new ErrorCode(INVALID_INPUT));
-        }
-
+        CertificateValidator.validateCertificate(giftCertificate);
         try {
             giftCertificate.setId(giftCertificateDAO.addGiftCertificate(giftCertificate));
```
Comment (churaalena) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:189]: Why do we set id into object and than get it ?

File: web/src/main/java/com.epam.esm/controller/CertificateController.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    @GetMapping("/certificates/{id}")
+    public GiftCertificate getGiftCertificate(@PathVariable int id) throws ServiceException {
+        return giftCertificateService.getGiftCertificate(id);
+    }
+
+    @PostMapping("/certificates")
+    public GiftCertificate addGiftCertificate(@RequestBody GiftCertificate giftCertificate) throws ServiceException {
+        giftCertificate.setId(giftCertificateService.addGiftCertificate(giftCertificate));
```
Comment (churaalena) [inline web/src/main/java/com.epam.esm/controller/CertificateController.java:40]: why no to do this in service?

File: web/src/main/java/com.epam.esm/controller/CertificateController.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    private final GiftCertificateService giftCertificateService;
+
+    @Autowired
+    public CertificateController(GiftCertificateService giftCertificateService) {
+        this.giftCertificateService = giftCertificateService;
+    }
+
+    @GetMapping("/certificates")
+    public List<GiftCertificate> getGiftCertificates(
+            @RequestBody(required = false) CertificateRequestBody request) throws ServiceException {
```
Comment (churaalena) [inline web/src/main/java/com.epam.esm/controller/CertificateController.java:25]: why not in service?

File: service/src/main/java/com/epam/esm/service/exception/ServiceException.java (hunk 4 above)
Comment (AlekseySayarkin) [reply service/src/main/java/com/epam/esm/service/exception/ServiceException.java:6]: To ignore stack trace, cause, suppressed and localized message  from json serialization

File: web/src/main/java/com.epam.esm/controller/CertificateController.java (hunk 11 above)
Comment (AlekseySayarkin) [reply web/src/main/java/com.epam.esm/controller/CertificateController.java:25]: done

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java (hunk 7 above)
Comment (AlekseySayarkin) [reply service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:92]: done

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java (hunk 5 above)
Comment (AlekseySayarkin) [reply service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:25]: changed to public

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java (hunk 6 above)
Comment (AlekseySayarkin) [reply service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:47]: Done

File: web/src/main/java/com.epam.esm/controller/CertificateController.java (hunk 10 above)
Comment (AlekseySayarkin) [reply web/src/main/java/com.epam.esm/controller/CertificateController.java:40]: Reworked to set id in service

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java (hunk 9 above)
Comment (AlekseySayarkin) [reply service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:189]: Introduced var id

**Pull request merged**